### PR TITLE
[439] Add contact details progress state

### PR DESCRIPTION
**Pull request by aliuk2012**

### CODE REVIEW

File: app/controllers/trainees/contact_details_controller.rb [hunk 1]
```diff
@@ -26,5 +30,20 @@ def contact_details_params
         *ContactDetail::FIELDS,
       )
     end
+
+    def contact_details_attributes
+      ContactDetails::Utilities::AddressAttributes.new(contact_details_params).attributes
```
Comment (jhenderson) [inline app/controllers/trainees/contact_details_controller.rb:35]: I suggest deleting `ContactDetails::Utilities::AddressAttributes` (it's not service object anyway) and put all the logic into the `ContactDetail` model.

File: app/services/contact_details/utilities/address_attributes.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+      end
+
+      def attributes
+        uk_locale? ? @attributes.merge(reset_non_uk_attributes) : @attributes.merge(reset_uk_attributes)
+      end
+
+    private
+
+      def uk_locale?
+        @attributes[:locale_code] == "uk"
```
Comment (jhenderson) [inline app/services/contact_details/utilities/address_attributes.rb:15]: There are several instances of the string "uk". We could replace these strings with a constant say `UK_LOCALE_CODE`

**Review by jhenderson — CHANGES_REQUESTED** [review]

File: app/controllers/trainees/contact_details_controller.rb (hunk 1 above)
Comment (aliuk2012) [reply app/controllers/trainees/contact_details_controller.rb:35]: @darokel ^

Comment (darokel) [reply app/controllers/trainees/contact_details_controller.rb:35]: I suggest keeping it out. The `ContactDetail` is really just a validator. Its job is to validate the attributes currently on a given trainee. Manipulating the attributes should be handled outside to keep this class lightweight and simple.

**Comment by aliuk2012** [comment]
[Contact details progress/state](https://trello.com/c/SNrHqtYI/439-contact-details-progress-state)

File: app/services/contact_details/utilities/address_attributes.rb (hunk 2 above)
Comment (aliuk2012) [reply app/services/contact_details/utilities/address_attributes.rb:15]: Well, the locale code is meant to be an enum. Could I suggest we do it as a separate ticket to move this to a constant?

**Review by darokel — APPROVED** [review]

File: app/controllers/trainees/contact_details_controller.rb (hunk 1 above)
Comment (jhenderson) [reply app/controllers/trainees/contact_details_controller.rb]: @darokel Check out the last commit. @aliuk2012 was helping me fix a frontend issue I was having earlier and then we had a chat about this particular area of code. We ended up pairing for a bit to see if there was another way. Let us know what you think.

Comment (darokel) [reply app/controllers/trainees/contact_details_controller.rb]: It's out of the validator so I'm happy with that. Tacking it on the trainee and having to execute the logic via an explicit method makes sense. I've already approved the PR so will leave it at that 👍

**Review by jhenderson — APPROVED** [review]

**Pull request merged**